### ISSUE TITLE
Results on cifar-10

### ISSUE THREAD

**Issue by jwyang** [rusty1s/pytorch_geometric#3]
Thanks for sharing this nice code!

I think it is a very promising direction. I am wondering whether you ever tried to apply splineConv to cifar-10, like that on mnist? It would be great to know whether it is also applicable to more complicated dataset? if any, what is the bottleneck to apply it to more complex image dataset, like ImageNet?

thanks again for such a great work!

Jianwei

**Comment by rusty1s** [comment]
Thanks Jianwei!

Actually, i applied a previous version of SplineConv on CIFAR-10 superpixels generated by Quickshift and SLIC ([you can find the code here](https://github.com/rusty1s/embedded_gcnn)). Note that the TensorFlow implementation is quite slow and the early theory behind SplineConv kinda lackluster. Generally speaking, it is not advisable to use Graph Convolutions for the task of like image classification (use conv2d for that) and instead use them for real graph problems.

However, it should be quite trivial to apply SplineCNN to more complicated image datasets:
* Use an existing PyTorch dataset like CIFAR-10 and create a transforms object, which takes images and generates superpixel graphs. You can then even apply augmentation on the images before generating superpixels.
* You can precompute superpixel graphs of images and save them in their own `*.pt` file (like in the MNISTSuperpixelsDataset). If the dataset does not fit into memory, it is advisable to generate a single graph file for a single image and load them just like in the normal ImageNet Dataset in PyTorch.

Bottlenecks may be the generation of superpixel graphs (e.g. the scikit segmentation algorithms are kinda slow) and the GraclusPooling (which should be precomputed on large datasets).

**Comment by jwyang** [comment]
Hi, Matthias

thanks a lot for your detailed explanation. I have a better sense now. Generating superpixels and perform GraclusPooling seems very time consuming when I run your code on mnist-superpixel. What if we can accelerate the computation, do you think it is possible to beat the conventional CNN, such as vgg on large-scale image classification, e.g., CIFAR-10 or ImageNet? I am asking because I intuitively think building the hierarchical graph representation of images like this might be better than grid-like representations. As a side-product, the model might also become more robust to adversarial attacks. What do you think?

**Comment by rusty1s** [comment]
Hi,

I'm working on a better implementation of Graclus without the need of preprocessing, which should be finished early next year. This should give a huge performance boost.

To your question:
It might be possible that graph representations of images can beat grid-like representation. It all depends on your input data and how granular you represent detailed information of images in your graph. In addition, you may need handcrafted input features that describe the shape of superpixels (which kinda sucks). Therefore, it might be even better to think of a complete different graph representation of an image.

A note on performance:
Note that runtime of classical CNNs depend on the number of pixels, whereas GraphCNNs depend on the number of edges! That makes it nearly impossible to beat performance of classical CNNs. You need a huge reduction in nodes and a really sparse adjacency matrix in order to accomplish that.

**Comment by jwyang** [comment]
Hi, Matthias

Appreciate much for your response. that's really impressive! Looking forward to your next version of Graclus pooling!

Yes, I totally agree with your comments on my question. It might be difficult now to think about a decent graph representation of images, but one intermediate graph the proposed architecture generate at some level should be semantic. if the graph node can represent objects or object parts at some level, that would be great.

You are right that CNN is coping with nodes, while graphCNNs tackle edges. Maybe we can use a similar pooling method like max/average pooling to pool the graph nodes, and them sparsity the edges layer by layer. In this sense, it might achieve comparable efficiency to conventional CNNs. For example, at the earlier layer, nodes are object parts, and them nodes become objects at higher layer, resulting in reduced graph nodes, finally, the graph is pooled to a single node. To complete this, I think a smart pooling and connection strategy are indispensable.

**Comment by rusty1s** [comment]
Hi,

of course, this would be really cool. You could start of with a grid-representation of the image and generate more irregular-structured data at deeper layers. You need some kind of learned pooling in order to do this, and therefore Graclus seems not the way to go here.
Learned Pooling on graphs with a connection strategy should be, however, still an open problem, because you need to somehow backpropagate gradients through the pooling layers. If you have any plans on working on this, I would be very interested in following your project.

**Comment by jwyang** [comment]
sure, I will let you know how I am going along this direction!

**Comment by rusty1s** [comment]
Hello Jianwei,

in the meanwhile I implemented a Graclus GPU algorithm and you can find it in the current master branch. Please let me know if this is useful for you :)

I'm closing this issue for now.

**Comment by tonyandsunny** [comment]
Hi， I use the embedded_gcnn code to segment the cifar-10 images, and transforming it to the pytorch_geometric dataset format. But the result of GCNConv in cifar-10 is about 0.50. I can not understand it. Did you do the cifar-10 experiment in the graph conv?

**Comment by rusty1s** [comment]
Yes, a few years ago in my master thesis. You can find it [here](https://github.com/rusty1s/deep-learning-on-graphs/raw/master/masterthesis/main.pdf). It is in german though, but the Tables should still be understandable. Table 6.9 on page 86 gives the results for the CIFAR-10 dataset. For `GCNConv`, I was able to achieve an accuracy of 54%.

**Comment by tonyandsunny** [comment]
Thank you very much!
